### PR TITLE
First version of congestion control

### PR DESCRIPTION
**Pull request by ianswett**

### CODE REVIEW

File: draft-ietf-quic-recovery.md [hunk 1]
```diff
@@ -202,8 +202,8 @@ kDefaultInitialRtt (default 100ms):
 
 ## Variables of interest
 
-We first describe the variables required to implement the loss detection
-mechanisms described in this section.
+Variables required to implement the congestion control mechanisms
+described in this section.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: "are described in this section"

File: draft-ietf-quic-recovery.md [hunk 2]
```diff
@@ -506,16 +506,123 @@ Pseudocode for DetectLostPackets follows:
...
+       foreach (packet in lost_packets)
+         sent_packets.remove(packet.packet_number)
 ~~~
 
 # Congestion Control
 
-(describe NewReno-style congestion control {{?RFC6582}} for QUIC.)
-(describe appropriate byte counting.)
-(define recovery based on packet numbers.)
+QUIC's congestion control is based on TCP New Reno{{?RFC6582}}
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: NewReno -- 1 word

File: draft-ietf-quic-recovery.md [hunk 3]
```diff
@@ -506,16 +506,123 @@ Pseudocode for DetectLostPackets follows:
...
+congestion control to determine the congestion window and pacing rate.
+
+## Slow Start
+
+QUIC uses a slow start approach where the congestion window is increased
+by the same number of bytes as are acknowledged while in slow start.
+QUIC begins every connection in slow start and exits slow start upon
+loss or when the min_rtt increases by more than 1/8th for a full round.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: remove the 1/8th RTT -- that's mixing hystart into this doc

File: draft-ietf-quic-recovery.md [hunk 4]
```diff
@@ -506,16 +506,123 @@ Pseudocode for DetectLostPackets follows:
...
+
+## Recovery
+
+Recovery is a period of time beginning with detection of a lost packet.
+It ends when all packets outstanding at the time recovery began have been
+acknowledged or lost. During recovery, the congestion window is not
+increased and the pacing gain is reduced to 1.
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I'd remove mention of "pacing gain"

File: draft-ietf-quic-recovery.md [hunk 5]
```diff
@@ -506,16 +506,123 @@ Pseudocode for DetectLostPackets follows:
...
+
+Invoked by loss detection from DetectLostPackets when new packets
+are detected lost.
+
+~~~
+   OnPacketsLost(lost_packets):
+     largest_lost_packet = lost_packets.last()
+     if (end_of_recovery < largest_lost_packet.packet_number):
+       end_of_recovery = largest_sent_packet
+       congestion_window *= kLossReductionFactor
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: ssthresh = congestion_window

File: draft-ietf-quic-recovery.md [hunk 6]
```diff
@@ -506,16 +506,123 @@ Pseudocode for DetectLostPackets follows:
...
+~~~
+
+## Sending Packets
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: I would rename this section  "Pacing Packets" or something like it

File: draft-ietf-quic-recovery.md [hunk 7]
```diff
@@ -506,16 +506,123 @@ Pseudocode for DetectLostPackets follows:
...
+
+~~~
+   TimeToSend(packet_size):
+     if (bytes_in_flight + packet_size > congestion_window)
+       return infinite
+     return time_of_last_sent_packet + 
+         packet_size * smoothed_rtt / congestion_window
+~~~
+
 (describe min_rtt based hystart.)
```
Comment (janaiyengar) [inline draft-ietf-quic-recovery.md]: Remove this line

File: draft-ietf-quic-recovery.md (hunk 1 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Done

File: draft-ietf-quic-recovery.md (hunk 2 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Done

File: draft-ietf-quic-recovery.md (hunk 3 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Oh, NewReno doesn't have hystart?  How 80s.

File: draft-ietf-quic-recovery.md (hunk 4 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Done.

File: draft-ietf-quic-recovery.md (hunk 5 above)
Comment (ianswett) [inline draft-ietf-quic-recovery.md]: Thanks, I forgot about ssthresh.

File: draft-ietf-quic-recovery.md [hunk 8]
```diff
@@ -506,17 +506,124 @@ Pseudocode for DetectLostPackets follows:
...
+kLossReductionFactor (default 0.5):
+: Reduction in congestion window when a new loss event is detected.
+
+
+## Variables of interest
+
+Variables required to implement the congestion control mechanisms
+are described in this section.
+
+bytes_in_flight:
```
Comment (mikkelfj) [inline draft-ietf-quic-recovery.md]: What exactly is this measuring: IP, UDP and QUIC header? ACK's, or just the sum of retransmittable frames, or just stream payload?

Comment (martinthomson) [inline draft-ietf-quic-recovery.md]: @mikkelfj, good question.  Could you open an issue?

Comment (mikkelfj) [inline draft-ietf-quic-recovery.md]: done

Comment (martinthomson) [inline draft-ietf-quic-recovery.md]: thanks

**Pull request merged**